### ISSUE TITLE
Unable to pull all service health

### ISSUE THREAD

**Issue by pete0emerson** [cablehead/python-consul#24]
With Consul.Health.service() I can pull the current health of a specified service. I'd like to be able to pull the current health of *all* services. This is how I do it with the consul CLI:

consul watch -type=checks cat

Like the current Health.service(), I need to use an index so that I can continually pull changes to the health checks. It looks like this is not possible with Consul.Health.service(), and I don't see how to do it with any of the other classes.

**Comment by pete0emerson** [comment]
Hmmm, I was hoping to be able to implement this myself and issue a pull request, but it's not clear which consul API endpoint I'd need to hit. Consul does it with their CLI, so I'll try to unwrap that a bit and post detail here.

**Comment by pete0emerson** [comment]
Aha!

The API endpoint is:

`/v1/health/state/any`

That, combined with the `X-Consul-Index` HTTP header, should be sufficient.

I'll code up a solution.

Thinking out loud, should all services be returned if the service passed to Consul.Health.service() is None, or should it be a new endpoint, (Consul.Health.services() ??).

**Comment by cablehead** [comment]
Hi Pete!

Yeah, it looks like you are after this endpoint: https://consul.io/docs/agent/http/health.html#health_state

It should be a new method, Health.state, which takes one argument that's either: ```any, unknown, passing, warning, or critical```

**Comment by cablehead** [comment]
Yeah, want to have a shot at coding it up? I can too if you like?

It should just come with a test as well:

https://github.com/cablehead/python-consul/blob/master/tests/test_std.py#L280

**Comment by pete0emerson** [comment]
I'll take a shot at it. Seems pretty straightforward. Do you like the "if service is None, return all services" approach?

**Comment by cablehead** [comment]
I'm just trying to keep python-consul as close to the raw consul API as possible. Also, I'm not sure with that approach what the api for state/passing should be, for instance.

**Comment by pete0emerson** [comment]
Oh, definitely.

Hmmm... state() function inside the Health class, perhaps? I'll monkey with it and see if I can figure something out that stays true to their API.

**Comment by cablehead** [comment]
Maybe: Health.state(name, index=None), for the signature? 

It should also take consistency and dc args too, but Health.service doesn't yet.. so maybe punt on that, and they could all be updated in one shot..

**Comment by pete0emerson** [comment]
Yup, that feels right to me. I'll do some coding and post for review once I have it.

**Comment by pete0emerson** [comment]
This seems to be working for me ... I'll try to write some tests for it next.

```python
        def state(self, name, index=None):
            """

            Returns a tuple of (*index*, *nodes*)

            *name* is a supported state. From the Consul docs:

                The supported states are any, unknown, passing, warning, or
                critical. The any state is a wildcard that can be used to return
                all checks.

            *index* is the current Consul index, suitable for making subsequent
            calls to wait for changes since this query was last run.

            *nodes* are the nodes providing the given service.

            """
            params = {}
            if index:
                params['index'] = index
            def callback(response):
                data = json.loads(response.body)
                return response.headers['X-Consul-Index'], data

            return self.agent.http.get(
                callback,
                '/v1/health/state/%s' % name, params=params)
```

**Comment by cablehead** [comment]
Awesome! Maybe just throw in an assert too? To check name is valid?

**Comment by pete0emerson** [comment]
Interesting thing about that assert, I thought about that too, but the API allows you to curl a non-existent state and returns a 200 OK:

```bash
$ curl -v localhost:8500/v1/health/state/fdshjfkdshfjsk
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8500 (#0)
> GET /v1/health/state/fdshjfkdshfjsk HTTP/1.1
> User-Agent: curl/7.37.1
> Host: localhost:8500
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Type: application/json
< X-Consul-Index: 1027
< X-Consul-Knownleader: true
< X-Consul-Lastcontact: 0
< Date: Wed, 04 Mar 2015 22:10:30 GMT
< Content-Length: 2
<
* Connection #0 to host localhost left intact
[]
```

Based on that, what would you prefer?

**Comment by cablehead** [comment]
Hrm, that is interesting. I think include the assert? Otherwise, a typo would go unnoticed. I'm sure Consul should return a 400 bad request there..

**Comment by cablehead** [comment]
https://github.com/hashicorp/consul/issues/757

**Comment by pete0emerson** [comment]
By the way, that's testing under 0.4.1, consul has released 0.5.0. I haven't tested it with 0.5.0, though.

**Comment by pete0emerson** [comment]
All righty, pull request is in: https://github.com/cablehead/python-consul/pull/25
